### PR TITLE
Create winget.yml

### PR DESCRIPTION
**Pull request by vedantmgoyal2009**
This action automatically generates manifests for WinGet Community Repository ([microsoft/winget-pkgs](https://github.com/microsoft/winget-pkgs)) and submits them.

Before merging this:
1. Please add a GitHub token with `public_repo` and `workflow` scopes as a repository secret and rename the secret name in the workflow.
2. Fork [microsoft/winget-pkgs](https://github.com/microsoft/winget-pkgs) under @OpenKneeboard.

### CODE REVIEW

**Comment by fredemmott** [comment]
Thanks! Is the workflow scope mandatory? I see you have auto-updating enabled, but in line with [GitHub's recommendations](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#reusing-third-party-workflows), I'd be using this with a full-size commit hash which I would manually update when needed - especially as there are other secrets in this repository.

**Comment by vedantmgoyal2009** [comment]
You can install https://github.com/apps/pull on https://github.com/OpenKneeboard/winget-pkgs. After this, you can omit the `workflow` scope and the action will still work.

**Comment by fredemmott** [comment]
Actually, sorry - I'm not going to be taking this, because it doesn't fit my *usual* workflow: I publish to winget immediately for small changes, but for large new releases, I want an intentional delay between publishing and it going out to auto-update.

That said, I greatly appreciate your efforts to make this easier both for me, and the winget community as a whole.

**Comment by vedantmgoyal2009** [comment]
I've also added https://github.com/OpenKneeboard/OpenKneeboard/network/updates which automatically updates action version. (https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuring-dependabot-version-updates#checking-the-status-of-version-updates)

**Comment by vedantmgoyal2009** [comment]
> I publish to winget immediately for small changes, but for large new releases, I want an intentional multi-day delay between publishing and it going out to auto-update.

Since you follow semantic versioning, we can check compare the previous & current version and if its a minor/patch update, the workflow will continue, otherwise it will abort.

**Comment by fredemmott** [comment]
Not strict semver, but yep, as long as it understands to ignore prerelease for that check, that’ll work - in that case it should only update for .z changes

**Comment by vedantmgoyal2009** [comment]
> it understands to ignore prerelease for that check

The existing workflow already ignores pre-release. When the pre-release will be converted to a release, it will still be able to run 😉

File: .github/workflows/winget.yml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+name: Publish to WinGet
+on:
+  release:
+    types: [released]
```
Comment (vedantmgoyal2009) [inline .github/workflows/winget.yml:4]: image

Comment (vedantmgoyal2009) [reply .github/workflows/winget.yml:4]: @fredemmott the above screenshot is from https://docs.github.com/webhooks-and-events/webhooks/webhook-events-and-payloads?actionType=released#release.

**Comment by fredemmott** [comment]
Sorry, I mean purely for checking what the change is, not for when it runs:

if I release v1.3.4, v1.4beta1 (prerelease), then v1.3.5 (release), the workflow condition needs to be comparing v1.3.4 to v1.3.5, not v1.4beta1 to v1.3.5

**Comment by fredemmott** [comment]
This also works on other events, as long as I specify the version, right?

if so, I’ll probably:
- take everything dependabot-related here verbatim
- add a workflow to run the scripts to update autoupdate/ for stable or preview as appropriate whenever any release/prerelease is made, and create a PR
- Run winget-release workflow whenever there’s a push to master that modifies autoupdate/stable.json

**Comment by vedantmgoyal2009** [comment]
> * Run winget-release workflow whenever there’s a push to master that modifies autoupdate/stable-msi.json

The action does not run on `push` events. Let me simplify the things:
- The workflow will run on every release or a pre-release is converted to a release
- Then, it will get the last "release" (not pre-release) tag, and compare it to current tag
  - If it is a semver major update, it will abort silently.
    - When the intentional multi-day delay is complete, you will have to trigger the workflow manually and specify the release tag you want to publish.
  - Else, it will publish to WinGet normally

**Comment by fredemmott** [comment]
Yes, I understand what the PR is doing, I was suggesting an alternative approach.

Is winget-releaser compatible with push events if the version is specified, regardless of what the yaml is currently specifying in this PR?

**Comment by fredemmott** [comment]
And the answer to that is yes, as shown by workflow_dispatch :)

**Comment by vedantmgoyal2009** [comment]
image

Here is an example of how you will be specifying the version when running on the `workflow_dispatch` event.

**Comment by fredemmott** [comment]
@vedantmgoyal2009 could you either:
- split out the dependabot changes to a separate PR
- change the winget workflow to:
  - only run on pushes that
    - are to OpenKneeboard/OpenKneeboard
    - are to master
    - touch `autoupdate\stable-msi.json`
  - retrieve the version with `(Get-Content .\autoupdate\stable-msi.json | ConvertFrom-Json).tag_name`
  - set `fork-user: openkneeboard-bot`
- ... or both :)

I'm also happy to do this myself, if you're fine with it being outside this PR (i.e. you would get commit attribution for cherry-picks, but github wouldn't consider it/credit it as a merged PR)

File: .github/workflows/winget.yml [hunk 2]
```diff
@@ -2,12 +2,50 @@ name: Publish to WinGet
...
+            Write-Error "Cannot determine tag to publish"
+            exit 1
+          }
+          
+          # Get the last tag ignoring pre-releases using the GitHub API
+          $lastTag = (Invoke-RestMethod -Uri "https://api.github.com/repos/${{ github.repository }}/releases/latest").tag_name
+
+          # Compare the last tag to the current tag, and only publish if it is semver minor/patch update
+          $lastVersion = [System.Version]::Parse($lastTag.Substring(1))
+          $currentVersion = [System.Version]::Parse($tag)
```
Comment (fredemmott) [inline .github/workflows/winget.yml]: FYI, `[System.Management.Automation.SemanticVersion]` is a more permissive parser, and required for this repo (which doesn't follow semver strictly)

**Comment by vedantmgoyal2009** [comment]
@fredemmott since the workflow will be using `tag_name` from the JSON file, should it still be comparing the semver? You already have full control over the releases.

File: .github/workflows/winget.yml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        shell: pwsh
+        id: should-publish
+
+      - if: ${{ steps.should-publish.outputs.publish == 'true' }}
+        uses: vedantmgoyal2009/winget-releaser@v2
+        with:
+          identifier: FredEmmott.OpenKneeboard
+          version: ${{ steps.should-publish.outputs.version }}
+          fork-user: openkneeboard-bot
+          token: ${{ secrets.OPENKNEEBOARD_BOT_PAT }}
```
Comment (fredemmott) [inline .github/workflows/winget.yml:34]: Thanks, well-spotted :)

File: .github/workflows/winget.yml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+jobs:
+  publish:
+    # Action can only be run on windows
+    runs-on: windows-latest
+    steps:
+      - name: Should publish to WinGet?
+        run: |-
+          # Get the tag from the autoupdate/stable-msi.json
+          $githubRawUrl = "https://raw.githubusercontent.com/OpenKneeboard/OpenKneeboard/master/autoupdate/stable-msi.json"
+          $tag = (Invoke-RestMethod -Uri $githubRawUrl).tag_name
```
Comment (fredemmott) [inline .github/workflows/winget.yml:18]: TIL Invoke-RestMethod deals with JSON, nice.

**Review by fredemmott — APPROVED** [review]

**Comment by fredemmott** [comment]
This looks good, but GitHub wants a manual rebase (perhaps a squash would be best)

**Comment by vedantmgoyal2009** [comment]
Squash will be good, otherwise, the repo will 7 extra commits which will make commit history dirty.

File: .github/workflows/winget.yml (hunk 4 above)
Comment (vedantmgoyal2009) [reply .github/workflows/winget.yml:18]: By the way, do you know PowerShell?

Comment (fredemmott) [reply .github/workflows/winget.yml:18]: Just what I've learned through github actions for windows/cross-platform stuff. Professionally, Linux/Mac exclusive, so still prefer POSIX shell for Linux/Mac-only stuff (I know powershell is available).

**Comment by vedantmgoyal2009** [comment]
@fredemmott if you haven't given the PAT `workflow` scope, remember to install https://github.com/apps/pull on the fork (https://github.com/openkneeboard-bot/winget-pkgs).

**Comment by fredemmott** [comment]
Thanks, that's already on there :)

**Comment by fredemmott** [comment]
@vedantmgoyal2009 FWIW, I strongly recommend moving over to by-hash with dependabot/pull without the workflow permissions as your default recommendation:

If your account ever gets compromised, the potential damage to the winget ecosystem (and beyond winget for every repository that grants a workflow scope) is massive: it allows exfilitrating all github action secrets for all repos that have granted workflow scope - e.g. other access tokens (both github and for external services), code signing private keys, etc.

While GitHub won't pass them to your workflow unless requested, the permission gives the ability to add new workflows that do request them.

**Comment by fredemmott** [comment]
For my reference, flow is:

1. I push a tag
2. https://github.com/OpenKneeboard/OpenKneeboard/blob/master/.github/workflows/ci.yml will build and sign the artifacts, and create a draft release
3. I edit the release notes then publish
4. https://github.com/OpenKneeboard/OpenKneeboard/blob/master/.github/workflows/update-autoupdate.yml creates a PR like https://github.com/OpenKneeboard/OpenKneeboard/pull/270
5. I merge the PR
6. for *stable* channel (unlike that PR, which is preview channel), https://github.com/OpenKneeboard/OpenKneeboard/blob/master/.github/workflows/winget.yml will then send it to winget

**Comment by vedantmgoyal2009** [comment]
I didn't know that dependabot supports updating commit hash until today I saw the PR on your repo. I'll add some documentation about using commit hash along with using dependabot (https://github.com/vedantmgoyal2009/winget-releaser/issues/104) when I get some time.

**Comment by fredemmott** [comment]
This appears to be ignoring the tag and is attempting to find a tag called 'master' instead.

https://github.com/OpenKneeboard/OpenKneeboard/actions/runs/4484622156/jobs/7885322389

**Comment by fredemmott** [comment]
@vedantmgoyal2009 thanks, yep - https://github.com/OpenKneeboard/OpenKneeboard/commit/2a9eafdd22223b53c272121bfa4e0d643a2c87c6 might be worth pre-emptively applying in any other repos where you're doing similar things - the old approach is due to stop working in may

**Pull request merged**